client/ext/bigquery/current-timestamp-optimizer.js

```diff
@@ -0,0 +1,186 @@
+/**
+ * @copyright Copyright 2016 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * @fileoverview The CurrentTimestampOptimizer replaces the non-cacheable
+ *    CURRENT_TIMESTAMP() in queries with a fixed datetime based on the specified
+ *     granularity.
+ * 
+ * Granularity is used to provide an estimate of the current date/time, rounded to the nearest
+ * previous 'x'.  An example of this can be seen below:
+ *     original    CURRENT_TIMESTAMP()  2016-02-13 14:43:26
+ *     granularity YEAR                 2016-01-01
+ *                 MONTH                2016-02-01
+ *                 WEEK                 2016-02-06 (Saturday of previous week)
+ *                 DAY                  2016-02-13
+ *                 HOUR                 2016-02-13 14:00:00
+ *
+ * @author joemu@google.com (Joe Allan Muharsky)
+ */
+
+goog.provide('p3rf.perfkit.explorer.ext.bigquery.CurrentTimestampGranularity');
+goog.provide('p3rf.perfkit.explorer.ext.bigquery.CurrentTimestampOptimizer');
+goog.provide('p3rf.perfkit.explorer.ext.bigquery.CurrentTimestampOptimizerConfigModel');
+
+
+goog.scope(function() {
+  const bigquery = p3rf.perfkit.explorer.ext.bigquery;
+
+  /**
+   * Constants describing the types of granularity supported on timestamps.
+   * @enum
+   */
+  bigquery.CurrentTimestampGranularity = {
+    YEAR: 'YEAR',
+    MONTH: 'MONTH',
+    WEEK: 'WEEK',
+    DAY: 'DAY',
+    HOUR: 'HOUR'
+  };
+  const CurrentTimestampGranularity = bigquery.CurrentTimestampGranularity;
+
+  /**
+   * Describes the properties for configuring a CurrentTimestampOptimizer.
+   *
+   * @constructor
+   */
+  bigquery.CurrentTimestampOptimizerConfigModel = class {
+    constructor(enabled=null, granularity=null) {
```

Comment:
This should be a separate discussion, ok to keep the usage as-is for now.

For background, my goal would be to be able to do interactive development with instant reloads without a separate compile step, including having Karma tests having working autorefresh that picks up changed code, and in my initial experiments this seemed doable. The two blockers were default arguments (due to Chrome's current ES6 support not implementing them), and a bug in Angular related to class constructors.